### PR TITLE
MM-19606- Rework Prepackaged Plugins

### PR DESCRIPTION
**Pull request by marianunez**
#### Summary

Prepackaged plugins will continue to be included in the `prepackaged_plugins` folder but will they will not be installed automatically and now they can be upgraded and uninstalled: 

- New `AutomaticPrepackagedPlugins` setting will allow to automatically install and upgrade (if available) at startup time any prepackaged plugin that has was previously enabled. If false or is an older version, it will ignore the bundle in the prepackaged folder.
- New `EnableRemoteMarketplace` setting will allow disabling fetching plugins from the marketplace server when false. 
- `GetMarketplacePlugins` will now include local results for prepackaged plugins and installed plugins not listed in the marketplace server.
- Prepackaged plugins updated to support signature verification. 

#### Related PR
WebApp: https://github.com/mattermost/mattermost-webapp/pull/4564

#### Ticket Link
[MM-19606](https://mattermost.atlassian.net/browse/MM-19606)

### CODE REVIEW

**Comment by marianunez** [comment]
Will add dev reviewers after the holiday break.

File: app/plugin.go [hunk 1]
```diff
@@ -423,8 +409,115 @@ func (a *App) GetPlugins() (*model.PluginsResponse, *model.AppError) {
 	return resp, nil
 }
 
-// GetMarketplacePlugin returns plugin from marketplace-server
-func (a *App) GetMarketplacePlugin(request *model.InstallMarketplacePluginRequest) (*model.BaseMarketplacePlugin, *model.AppError) {
+func (a *App) InstallMarketplacePlugin(request *model.InstallMarketplacePluginRequest) (*model.Manifest, *model.AppError) {
```
Comment (marianunez) [inline app/plugin.go]: @ali-farooq0 Should we move this to `plugin_install.go` to keep all app installation code there?

Comment (ali-farooq0) [reply app/plugin.go]: Good idea @marianunez. Makes sense to keep all the installation code in `plugin_install.go` 👍

File: app/plugin.go [hunk 2]
```diff
@@ -618,5 +755,112 @@ func (a *App) getPluginsFromFolder() (map[string]*pluginSignaturePath, *model.Ap
...
+		mlog.Error("Failed to walk prepackaged plugins", mlog.Err(err))
+		return nil
+	}
+
+	pluginSignaturePathMap := getPluginsFromFilePaths(fileStorePaths)
+	plugins := make([]*plugin.PrepackagedPlugin, 0, len(pluginSignaturePathMap))
+	for _, pluginPaths := range pluginSignaturePathMap {
+		plugin, err := a.processPrepackagedPlugin(pluginPaths)
+		if err != nil {
+			mlog.Error("Failed to install prepackaged plugin %s", mlog.Err(err), mlog.String("path", pluginPaths.path))
```
Comment (agnivade) [inline app/plugin.go]: Are you sure the `%s` gets replaced ? From what I see of the method signature, it is a message with a slice of fields.

File: app/plugin.go [hunk 3]
```diff
@@ -465,55 +557,96 @@ func (a *App) GetMarketplacePlugins(filter *model.MarketplacePluginFilter) ([]*m
...
+		}
+
+		result[p.Manifest.Id] = &model.MarketplacePlugin{BaseMarketplacePlugin: p}
+	}
+
+	return result, nil
+}
+
+// mergePrepackagePlugins merges pre-packaged plugins to remote marketplace plugins list.
+func (a *App) mergePrepackagePlugins(remoteMarketplacePlugins map[string]*model.MarketplacePlugin) *model.AppError {
```
Comment (agnivade) [inline app/plugin.go]: Let's add some tests for mergePrepackagePlugins and mergeLocalPlugins ? I didn't see them.

File: api4/plugin.go [hunk 4]
```diff
@@ -347,45 +326,17 @@ func parseMarketplacePluginFilter(u *url.URL) (*model.MarketplacePluginFilter, e
 
 	filter := u.Query().Get("filter")
 	serverVersion := u.Query().Get("server_version")
+	localOnly := u.Query().Get("local_only") == "true"
```
[the remark this answers is not in the document]
Comment (marianunez) [reply api4/plugin.go]: Good catch. Updated!

We currently have a mix of this and this more correct approach with `strconv.ParseBool`. HW was created to standardize across the API https://github.com/mattermost/mattermost-server/issues/13597

File: api4/plugin_test.go [hunk 5]
```diff
@@ -767,6 +774,273 @@ func TestSearchGetMarketplacePlugins(t *testing.T) {
...
+		})
+
+		sort.SliceStable(expectedPlugins, func(i, j int) bool {
+			return strings.ToLower(expectedPlugins[i].Manifest.Id) < strings.ToLower(expectedPlugins[j].Manifest.Id)
+		})
+		sort.SliceStable(plugins, func(i, j int) bool {
+			return strings.ToLower(plugins[i].Manifest.Id) < strings.ToLower(plugins[j].Manifest.Id)
+		})
+
+		require.EqualValues(t, expectedPlugins, plugins)
```
[the remark this answers is not in the document]
Comment (marianunez) [reply api4/plugin_test.go]: Thank you for this! I kept looking for something to ignore the order but couldn't find it (still ramping up on Go here 🙋‍♀️)

File: api4/plugin_test.go [hunk 6]
```diff
@@ -957,6 +1236,221 @@ func TestInstallMarketplacePlugin(t *testing.T) {
...
+		manifest2, resp := th.SystemAdminClient.InstallMarketplacePlugin(pRequest)
+		CheckNoError(t, resp)
+		require.NotNil(t, manifest2)
+		require.Equal(t, "testplugin2", manifest2.Id)
+		require.Equal(t, "1.2.3", manifest2.Version)
+
+		pluginsResp, resp = th.SystemAdminClient.GetPlugins()
+		CheckNoError(t, resp)
+		require.Len(t, pluginsResp.Active, 0)
+		require.Equal(t, pluginsResp.Inactive, []*model.PluginInfo{
```
[the remark this answers is not in the document]
Comment (marianunez) [reply api4/plugin_test.go]: I guess there is no real guarantee here. Replacing with the `require.ElementsMatch` 🎉

**Comment by hmhealey** [comment]
Just to confirm, is the NPS plugin still automatically enabled with this? I can't remember where the special handling for it lived.

**Comment by marianunez** [comment]
> Just to confirm, is the NPS plugin still automatically enabled with this? I can't remember where the special handling for it lived.

I believe the NPS plugins is still enabled by default on a fresh installation ( is this correct @lieut-data ?). However, if the admin has disabled it, it will be uninstalled on upgrade and be available to reinstall in the Plugins Marketplace. 

If it was still enabled, it will remain the same but now they can explicitly uninstall it like any other plugin (and also still available to reinstall from the marketplace).

File: app/plugin.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (marianunez) [reply app/plugin.go]: Done. Also moved it to the `plugin_install.go` as discussed here: https://github.com/mattermost/mattermost-server/pull/13449#discussion_r360623116

File: app/plugin.go [hunk 7]
```diff
@@ -618,5 +755,112 @@ func (a *App) getPluginsFromFolder() (map[string]*pluginSignaturePath, *model.Ap
...
+	prepackagedPluginsDir, found := fileutils.FindDir(pluginsDir)
+	if !found {
+		return nil
+	}
+
+	fileStorePaths := []string{}
+	if err := filepath.Walk(prepackagedPluginsDir, func(walkPath string, info os.FileInfo, err error) error {
+		fileStorePaths = append(fileStorePaths, walkPath)
+		return nil
+	}); err != nil {
```
[the remark this answers is not in the document]
Comment (marianunez) [reply app/plugin.go]: It seems that the `gofmt` in my editor does not like the new line there - it keeps removing it 🤷‍♀️

**Comment by lieut-data** [comment]
@marianunez has summarized correctly -- no changes to NPS plugin being auto-installed and enabled on startup. @prapti, let's make sure we explicitly test that as part of the test plan for this feature.

File: app/plugin.go [hunk 8]
```diff
@@ -4,12 +4,19 @@
...
+	"io"
+	"io/ioutil"
 	"net/http"
 	"os"
 	"path/filepath"
 	"sort"
 	"strings"
 
+	"github.com/blang/semver"
+	svg "github.com/h2non/go-is-svg"
```
[the remark this answers is not in the document]
Comment (marianunez) [reply app/plugin.go]: Moved to a separate group after the MM imports.

File: app/download.go [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+	"github.com/pkg/errors"
+)
+
+const (
+	// HTTP_REQUEST_TIMEOUT defines a high timeout for downloading large files
+	// from an external URL to avoid slow connections from failing to install.
+	HTTP_REQUEST_TIMEOUT = 1 * time.Hour
+)
+
+func (a *App) DownloadFromUrl(downloadUrl string) (io.ReadSeeker, error) {
```
[the remark this answers is not in the document]
Comment (marianunez) [reply app/download.go]: Done! https://github.com/mattermost/mattermost-server/pull/13449/commits/855c0bdaeed0474666c15787934b97d01a936d34

File: app/plugin.go (hunk 2 above)
Comment (marianunez) [reply app/plugin.go]: Good catch! With the structured logging, the placeholder is not replaced/needed. Updated.

File: app/plugin.go (hunk 3 above)
Comment (marianunez) [reply app/plugin.go]: They were tested as a whole with all the test cases of GetMarketplacePlugins in `api4/plugin_test`:
https://github.com/mattermost/mattermost-server/blob/855c0bdaeed0474666c15787934b97d01a936d34/api4/plugin_test.go#L789
https://github.com/mattermost/mattermost-server/blob/855c0bdaeed0474666c15787934b97d01a936d34/api4/plugin_test.go#L933
https://github.com/mattermost/mattermost-server/blob/855c0bdaeed0474666c15787934b97d01a936d34/api4/plugin_test.go#L462
https://github.com/mattermost/mattermost-server/blob/855c0bdaeed0474666c15787934b97d01a936d34/api4/plugin_test.go#L789

File: plugin/environment.go [hunk 10]
```diff
@@ -87,6 +97,10 @@ func (env *Environment) Available() ([]*model.BundleInfo, error) {
 	return scanSearchPath(env.pluginDir)
 }
 
+func (env *Environment) PrepackagedPlugins() []*PrepackagedPlugin {
+	return env.prepackagedPlugins
```
[the remark this answers is not in the document]
Comment (marianunez) [reply plugin/environment.go]: For 1- I've added the read lock. 
I see your point in 2, however this seems to be a general problem across the code base out of scope of this PR. I added a GoDoc to indicate that the list content should not be modified.

File: app/plugin.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (agnivade) [reply app/plugin.go]: Ah no .. I mean split the err into a separate line. If you just move the `err != nil` check, then gofmt will move it up ofcourse.

I meant:
```
err := filepath.Walk(...)
if err != nil {
}
```

[the remark this answers is not in the document]
Comment (marianunez) [reply app/plugin.go]: Gotcha!

**Comment by marianunez** [comment]
@lieut-data I've added the missing condition we discussed to skip installing the prepackaged plugin if there were not previously enabled here https://github.com/mattermost/mattermost-server/pull/13449/commits/d991135e32354c5dab1162570f6368bda4ba3217 if you want to take a look 👀

File: app/plugin.go [hunk 11]
```diff
@@ -756,8 +758,11 @@ func (a *App) processPrepackagedPlugin(pluginPath *pluginSignaturePath) (*plugin
 		return nil, errors.Wrapf(err, "Failed to get prepackaged plugin %s", pluginPath.path)
 	}
 
-	// Skip installing the plugin at all if automatic prepackaged plugins is disabled.
-	if !*a.Config().PluginSettings.AutomaticPrepackagedPlugins {
+	pluginsEnvironment := a.GetPluginsEnvironment()
+
+	// Skip installing the plugin at all if automatic prepackaged plugins is disabled
+	// or if the plugin is has not been enabled.
+	if !(*a.Config().PluginSettings.AutomaticPrepackagedPlugins && pluginsEnvironment.IsActive(plugin.Manifest.Id)) {
```
Comment (lieut-data) [inline app/plugin.go]: Two comments:
* I'm not sure `IsActive` is the right bit to check here. I think that will be true only if the plugin is installed and running. I think we need to check the configuration here, and determine if the plugin has been enabled or not.
* I personally find `!(condition && condition)` hard to parse: can we break these out explicitly instead?

Comment (marianunez) [reply app/plugin.go]: I've updated to use `PluginSettings.PluginStates` from the config.

File: app/plugin_install.go [hunk 12]
```diff
@@ -169,6 +172,59 @@ func (a *App) installPlugin(pluginFile, signature io.ReadSeeker, installationStr
...
+// from the prepackaged folder, if available, or remotely if EnableRemoteMarketplace is true.
+func (a *App) InstallMarketplacePlugin(request *model.InstallMarketplacePluginRequest) (*model.Manifest, *model.AppError) {
+	var pluginFile, signatureFile io.ReadSeeker
+
+	prepackagedPlugin, appErr := a.getPrepackagedPlugin(request.Id, request.Version)
+	if appErr != nil && appErr.Id != "app.plugin.marketplace_plugins.not_found.app_error" {
+		return nil, appErr
+	}
+	if prepackagedPlugin != nil {
+		fileReader, err := os.Open(prepackagedPlugin.Path)
```
Comment (crspeller) [inline app/plugin_install.go]: Close this file?

File: app/plugin_install.go [hunk 13]
```diff
@@ -169,6 +172,59 @@ func (a *App) installPlugin(pluginFile, signature io.ReadSeeker, installationStr
...
+		if err != nil {
+			err = errors.Wrapf(err, "failed to open prepackaged plugin %s", prepackagedPlugin.Path)
+			return nil, model.NewAppError("InstallMarketplacePlugin", "app.plugin.install_marketplace_plugin.app_error", nil, err.Error(), http.StatusInternalServerError)
+		}
+		pluginFile = fileReader
+		signatureFile = bytes.NewReader(prepackagedPlugin.Signature)
+	}
+
+	if *a.Config().PluginSettings.EnableRemoteMarketplace && pluginFile == nil {
```
Comment (crspeller) [inline app/plugin_install.go]: Isn't the remote marketplace supposed to override prepackaged if available? Maybe I don't understand how that is happening.

Comment (marianunez) [reply app/plugin_install.go]: It will override it if it has a newer version (and if the remote markerplace is allowed). The marketplace lists all latest versions available and when the installation is requested is done by version. If the prepackaged plugin has the latest version there is no need to fetch it from the remote source.

It will look first for the plugin in the prepackaged folder [here](https://github.com/mattermost/mattermost-server/blob/8673fcd9c918c5811baf4cd367cfc7ccd4a0b887/app/plugin_install.go#L180)(line 180 above) specifying the version requested. If it can't find the same version it then here it will go to the marketplace to get that newer version.

File: app/plugin_install.go (hunk 12 above)
Comment (marianunez) [reply app/plugin_install.go]: Yes, good catch!

File: app/plugin_install.go (hunk 13 above)
Comment (crspeller) [reply app/plugin_install.go]: Ah. So the version has already been determined at this point. I see.

**Comment by hanzei** [comment]
I've added the small change to only show the `Local` label if the remove marketplace is enabled. See https://github.com/mattermost/mattermost-server/pull/13449/commits/1035471dd138bc9865f367a4d936236e8f86fefa
